### PR TITLE
Collection literals: output type inferences

### PR DESCRIPTION
**Pull request by cston**
Add rules for *collection literal expressions* to *input type inference* and *output type inference* sections.

### CODE REVIEW

File: proposals/collection-literals.md [hunk 1]
```diff
@@ -317,6 +317,13 @@ A *collection literal inference*  is made *from* an expression `E` *to* a type `
   - Otherwise, if `Eᵢ` has type `Uᵢ` then a *lower-bound inference* is made *from* `Uᵢ` *to* `Tₑ`.
   - Otherwise, no inferences are made for `Eᵢ`.
 
+> 11.6.3.7 Output type inferences
+> 
+> An *output type inference* is made *from* an expression `E` *to* a type `T` in the following way:
+> 
+> - **If `E` is a *collection literal* with elements `Eᵢ` and `T` is a [*collection type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) with an [*iteration type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) `Tₑ`, then an *output type inference* is made *from* each `Eᵢ` *to* `Tₑ`.**
```
Comment (RikkiGibson) [inline proposals/collection-literals.md:324]: I didn't understand why this part is necessary. I thought the language earlier in this section (L314-L318) covers the same thing.

Comment (RikkiGibson) [reply proposals/collection-literals.md:324]: Maybe for stuff like `M(([1], ["a"]))`?

File: proposals/collection-literals.md [hunk 2]
```diff
@@ -317,6 +317,13 @@ A *collection literal inference*  is made *from* an expression `E` *to* a type `
   - Otherwise, if `Eᵢ` has type `Uᵢ` then a *lower-bound inference* is made *from* `Uᵢ` *to* `Tₑ`.
   - Otherwise, no inferences are made for `Eᵢ`.
 
+> 11.6.3.7 Output type inferences
```
[the remark this answers is not in the document]
Comment (cston) [reply proposals/collection-literals.md:320]: The links here are to standard-v6 currently. We could update all to v7 in the future and update the section numbers at the same time.

File: proposals/collection-literals.md (hunk 1 above)
Comment (cston) [reply proposals/collection-literals.md:324]: The *output type inference* change is to walk into elements for inferring from output types of elements, such as the lambda return type in:
```
F([null, () => 1]);

static void F<T>(Func<T>[] arg) { }
```

File: proposals/collection-literals.md [hunk 3]
```diff
@@ -302,20 +302,26 @@ static T[] AsArray<T>(T[] arg) => arg;
...
 
-The [*type inference*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#1163-type-inference) rules are **updated** to include inferences from collection literal expressions.
+The [*type inference*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#1163-type-inference) rules are updated as follows.
+
+The existing rules for the [*first phase*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#11632-the-first-phase) are extracted to a new *input type inference* section, and a  rule is added to *input type inference* and *output type inference* for collection literal expressions.
 
 > 11.6.3.2 The first phase
 > 
 > For each of the method arguments `Eᵢ`:
+> - An *input type inference* is made *from* `Eᵢ` *to* the corresponding *parameter type* `Tᵢ`.
```
Comment (cston) [inline proposals/collection-literals.md:312]: Extracting existing rules to a new *input type inference* based on feedback from @MadsTorgersen.

File: proposals/collection-literals.md [hunk 4]
```diff
@@ -302,20 +302,26 @@ static T[] AsArray<T>(T[] arg) => arg;
...
+
+> An *input type inference* is made *from* an expression `E` *to* a type `T` in the following way:
+>
+> - If `E` is a *collection literal* with elements `Eᵢ` and `T` is a [*collection type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) with an [*iteration type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) `Tₑ`, then an *input type inference* is made *from* each `Eᵢ` *to* `Tₑ`.
```
Comment (333fred) [inline proposals/collection-literals.md:316]: I'm a little concerned about the interaction of this rule with tuples. I think what the LDM wants is multi-level walks here, where a tuple of collection literals would walk through multiple levels if `T` is such a type, but this wouldn't do that.

Comment (cston) [reply proposals/collection-literals.md:316]: The rule for collection literals and the rule for tuple literals both use this *input type inference* rule for their parts.

File: proposals/collection-literals.md [hunk 5]
```diff
@@ -302,27 +302,26 @@ static T[] AsArray<T>(T[] arg) => arg;
 static List<T[]> AsListOfArray<T>(List<T[]> arg) => arg;
 ```
 
-The [*type inference*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#1163-type-inference) rules are **updated** to include inferences from collection literal expressions.
+The [*type inference*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#1163-type-inference) rules are updated as follows.
+
+The existing rules for the [*first phase*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#11632-the-first-phase) are extracted to a new *input type inference* section, and a  rule is added to *input type inference* and *output type inference* for collection literal expressions.
```
Comment (RikkiGibson) [inline proposals/collection-literals.md:307]:
```suggestion
The existing rules for the [*first phase*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#11632-the-first-phase) are extracted to a new *input type inference* section, and a rule is added to *input type inference* and *output type inference* for collection literal expressions.
```

**Review by RikkiGibson — APPROVED** [review]

**Review by 333fred — APPROVED** [review]

**Pull request merged**